Refactor: Test Framework and Update Docker Compose Configuration

rohansx opened the pull request:
Key Changes:
- Created a new `docker-compose.yml` file to manage the Xline cluster configuration, ensuring all services are correctly networked and equipped with the necessary volumes, ports, and environment settings.
- Refactored the `validation.sh` script to utilize Docker Compose for starting and stopping services, aligning it with the new docker-compose.yml configuration. This update simplifies the validation process and ensures compatibility with the Docker Compose setup.
- Updated the `benchmark.sh` and `quick_start.sh` documentation to reflect the new Docker Compose-based approach. This includes instructions on how to use the docker-compose file for setting up a development or testing environment and guidelines for executing validation and benchmark tests.

These changes aim to streamline the project's testing framework, reduce redundancy, and improve the overall developer experience when testing and validating the Xline cluster's functionality.

Resolves: #692

Phoenix500526 committed:
chore: mount LOG_PATH as xline container log dir

Signed-off-by: Phoeniix Zhao <Phoenix500526@163.com>

ChaudharyRaman committed:
fix: quick-start readme corrected pem file path

Signed-off-by: ChaudharyRaman <ramanchaudhary2704@gmail.com>
Signed-off-by: rohansx <sharmarohan615@gmail.com>

rohansx committed:
[200~Refactor Test Framework and Update Docker Compose Configuration

rohansx committed:
[200~Refactor Test Framework and Update Docker Compose Configuration

Signed-off-by: rohansx <sharmarohan615@gmail.com>

rohansx committed:
Merge branch 'master' of https://github.com/rohansx/Xline

rohansx committed:
Refactor: Test Framework and Update Docker Compose Configuration

rohansx committed:
Merge branches 'master' and 'master' of https://github.com/rohansx/Xline

rohansx committed:
fix: trailing space and bug

rohansx committed:
fix: trailing space and bug

rohansx committed:
fix: trailing space and bug

Signed-off-by: rohansx <sharmarohan615@gmail.com>

rohansx committed:
fix: trailing space and bug

Signed-off-by: rohansx <sharmarohan615@gmail.com>

rohansx committed:
feat: test framework and update docker compose configuration

Signed-off-by: rohansx <sharmarohan615@gmail.com>

rohansx commented:
@Phoenix500526 Pleaase review this PR and let me know the changes and then I'll proceed with the documentation

liangyuanpeng commented on ci/docker-compose.yml:
```diff
@@ -0,0 +1,69 @@
+    volumes:
+      - ./fixtures:/mnt
+    ports:
+      - "2380:2379"
```
I don't recommend using `2380:2379` here, because 2380 is a special port number, maybe `2479:2379`

liangyuanpeng replied:
@rohansx Maybe you missed here, there are no change,no?

liangyuanpeng commented on ci/docker-compose.yml:
```diff
@@ -0,0 +1,69 @@
+    volumes:
+      - ./fixtures:/mnt
+    ports:
+      - "2381:2379"
```
Same as above, maybe `2579:2379`

Phoenix500526 commented on ci/docker-compose.yml:
```diff
@@ -0,0 +1,69 @@
+    environment:
+      RUST_LOG: curp=debug,xline=debug
+    command: >
+      xline
```
Please add `--client-listen-urls` and `--peer-listen-urls` in the xline startup command.

Phoenix500526 commented on ci/docker-compose.yml:
```diff
@@ -0,0 +1,69 @@
+      --data-dir /usr/local/xline/data-dir
+      --auth-public-key /mnt/public.pem
+      --auth-private-key /mnt/private.pem
+
```
Here is missing the client container.

rohansx committed:
fix: adjust port configurations for xLine nodes and add client container for interaction

Signed-off-by: rohansx <sharmarohan615@gmail.com>

rohansx committed:
Merge branch 'xline-kv:master' into master

rohansx committed:
Merge branch 'test-framework'

rohansx commented:
@liangyuanpeng adjusted port configurations for xLine nodes and added client container for enhanced interaction.  @Phoenix500526 updated node2 and node3 ports to avoid special port conflicts and included --client-listen-urls and --peer-listen-urls for clearer networking. Also added a client service for testing and interaction purposes.

Phoenix500526 commented on ci/docker-compose.yml:
```diff
@@ -0,0 +1,85 @@
+      --peer-listen-urls http://172.22.0.4:2380
+
+  client:
+    image: appropriate/curl # Using a lightweight client for demonstration purposes
```
It should be ghcr.io/xline-kv/etcdctl:v3.5.9. This client container is used in the validation CI process. Please make the modification.

rohansx replied:
ok.. i will rebase this pr and make the modification.

Phoenix500526 reviewed (commented):
Please rebase this pr.

rohansx committed:
fix: update client container image for CI validation process

Signed-off-by: rohansx <sharmarohan615@gmail.com>

rohansx commented:
@Phoenix500526 I have done the necessary modification as you mentioned above. I updated client container image for CI validation process to ghcr.io/xline-kv/etcdctl:v3.5.9

Phoenix500526 commented:
There are still some problems with the validation CI. Please fix them so that this pr can pass the CI. You can run the validation on your local via executing `scripts/validation_test.sh`.

rohansx commented:
> There are still some problems with the validation CI. Please fix them so that this pr can pass the CI. You can run the validation on your local via executing `scripts/validation_test.sh`.

working on this.

rohansx committed:
fix: fix port number for node2

rohansx committed:
fix: fix port number for node3

rohansx committed:
Merge branch 'xline-kv:master' into master

rohansx committed:
fix: fix port number for node2

Signed-off-by: rohansx <sharmarohan615@gmail.com>

Phoenix500526 commented:
There are several issues in this pr: 
1. please add a signed-off-by information in your commits.
2. please format the commit message. The commit message "[200~Refactor Test Framework and Update..." is invalid.
3. The validation CI is still not pass. It seems the relevant container created by docker-compose named "ci-client-1", not `client`.

rohansx committed:
Merge branch 'xline-kv:master' into master

rohansx commented:
@Phoenix500526  @liangyuanpeng , I am encountering a validation check error, and I have tried to resolve it without success. Could you please explain why this validation check is failing and provide guidance on how I can address the issue?

Phoenix500526 commented:
> @Phoenix500526 @liangyuanpeng , I am encountering a validation check error, and I have tried to resolve it without success. Could you please explain why this validation check is failing and provide guidance on how I can address the issue?

I think there are several reasons behind this validation failure: 
1. The docker-compose will generate a default name for those containers without specifying their [container_name](https://docs.docker.com/compose/compose-file/compose-file-v3/#container_name). In this case, the name of the client container is `ci-client-1`, not `client`. Hence, the validation test will get an error like "No such container named client"
2. In this pr, you changed the subnet from `172.20.0.0/16` to `172.22.0.0/16`, but you forgot to update the ip address in the `validation_test.sh`. 
3. The original `quick_start.sh` mounts `./scripts` to `/mnt` and the `validation_test.sh` uses the `./scripts/validation_lock_client` to do some test stuff. 

Finally, please modify our [quick start doc](https://github.com/xline-kv/Xline/blob/master/doc/quick-start/README.md) to make it unify with the code change in this pr. Thanks 😄

Phoenix500526 commented:
Hi, @rohansx ! This pr has been stalled for 3 weeks. Would you like to update it? 😄

rohansx commented:
> Hi, @rohansx ! This pr has been stalled for 3 weeks. Would you like to update it? 😄

@Phoenix500526 yes I'll update in the next week.. sorry for the delay.. actually got stuck in the interview process and therefore i couldn't look into it..

Phoenix500526 commented:
@Mergifyio rebase
